src/Cardano/Wallet/DB/MVar.hs

```diff
@@ -33,36 +39,97 @@ import Data.Map.Strict
 
 import qualified Data.Map.Strict as Map
 
+data Database s = Database
+    { wallet :: Wallet s
+    , metadata :: WalletMetadata
+    , txHistory :: Map (Hash "Tx") (Tx, TxMeta)
+    }
+
 -- | Instantiate a new in-memory "database" layer that simply stores data in
 -- a local MVar. Data vanishes if the software is shut down.
 newDBLayer :: forall s. IO (DBLayer IO s)
 newDBLayer = do
-    db <- newMVar (mempty
-        :: Map (PrimaryKey WalletId) (Wallet s, Map (Hash "Tx") (Tx, TxMeta)))
+    db <- newMVar (mempty :: Map (PrimaryKey WalletId) (Database s))
     return $ DBLayer
-        { putCheckpoint = \key cp ->
-            let
-                alter = \case
-                    Nothing -> Just (cp, mempty)
-                    Just (_, history) -> Just (cp, history)
-            in
-                cp `deepseq` modifyMVar_ db (return . (Map.alter alter key))
+        {-----------------------------------------------------------------------
+                                      Wallets
+        -----------------------------------------------------------------------}
 
-        , readCheckpoint = \key ->
-            fmap fst . Map.lookup key <$> readMVar db
+        { createWallet = \key@(PrimaryKey wid) cp meta -> ExceptT $ do
+            let alter = \case
+                    Nothing ->
+                        Right $ Database cp meta mempty
+                    Just _ ->
+                        Left (ErrWalletAlreadyExists wid)
+            cp `deepseq` meta `deepseq` alterMVar db alter key
 
-        , readWallets =
+        , listWallets =
             Map.keys <$> readMVar db
 
+        {-----------------------------------------------------------------------
+                                    Checkpoints
+        -----------------------------------------------------------------------}
+
+        , putCheckpoint = \key@(PrimaryKey wid) cp -> ExceptT $ do
+            let alter = \case
+                    Nothing ->
+                        Left (ErrNoSuchWallet wid)
+                    Just (Database _ meta history) ->
+                        Right $ Database cp meta history
+            cp `deepseq` alterMVar db alter key
+
+        , readCheckpoint = \key ->
+            fmap wallet . Map.lookup key <$> readMVar db
+
+        {-----------------------------------------------------------------------
+                                   Wallet Metadata
+        -----------------------------------------------------------------------}
+
+        , putWalletMeta = \key@(PrimaryKey wid) meta -> ExceptT $ do
```

Comment:
wonder if fine-tuning type with "Symbol" would not do the trick here to embrace one impl of `putCheckpoint` and `readCheckpoint` for both checkpoint and metadata - the implementations are very close ....